### PR TITLE
feat: Eventing support for flagd provider

### PR DESCRIPTION
**Pull request by Kavindu-Dodan**
## This PR

Adds OpenFeature Events extension [^1] support to flagd provider

**tldr**

- Use`IService.go` as the main interface between provider and grpc handling (flagd connectivity)
- Cache is now extracted to a dedicated service (potential improvements exist here)
- Provider configurations are extracted to a dedicated component named `configuration.go`
- Eventing support 

Review commit by commit OR with a checkout of the original branch to have more focus 

----

Main refactoring on this PR focuses on isolating grpc based flagd connectivity to the service layer. This was done by refactoring `provider.go` and updating `IService.go`. With the change, we no longer have grpc dependencies at provider level. This is favorable for future efforts such as in-memory provider capability proposed through https://github.com/open-feature/flagd/pull/848 

Caching is now isolated to a dedicated service. This isolation helps to pass the service to appropriate caching locations and maintain a single cache within the provider. 

Provider configurations were also extracted to a new component to reduce logic at the provider level. 

Given below is an overview of component relationships in the current state, 

image

[^1]: https://openfeature.dev/specification/sections/events

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,7 +12,7 @@ test:
 	go list -f '{{.Dir}}/...' -m | xargs -I{} go test -v {}
 
 lint: 
-	go install -v github.com/golangci/golangci-lint/cmd/golangci-lint@latest
+	go install -v github.com/golangci/golangci-lint/cmd/golangci-lint@v1.54.1
```
Comment (Kavindu-Dodan) [inline Makefile:15]: the latest release is not compatible with go version 1.19+ variants

File: providers/flagd/README.md [hunk 2]
```diff
@@ -44,16 +44,16 @@ func main() {
...
-| maxCacheSize          | FLAGD_MAX_CACHE_SIZE           | int       |              | 1000      |
-| maxEventStreamRetries | FLAGD_MAX_EVENT_STREAM_RETRIES | int       |              | 5         |
+| Option name           | Environment variable name      | Type    | Options            | Default   |
+|-----------------------|--------------------------------|---------|--------------------|-----------|
+| host                  | FLAGD_HOST                     | string  |                    | localhost |
+| port                  | FLAGD_PORT                     | number  |                    | 8013      |
+| tls                   | FLAGD_TLS                      | boolean |                    | false     |
+| socketPath            | FLAGD_SOCKET_PATH              | string  |                    |           |
+| certPath              | FLAGD_SERVER_CERT_PATH         | string  |                    |           |
+| cache                 | FLAGD_CACHE                    | string  | lru, mem, disabled | lru       |
```
Comment (toddbaert) [inline providers/flagd/README.md:54]: Can you explain why you added this new cache type?

I guess it's not a problem, but I wonder why anyone would use it over LRU.

File: providers/flagd/pkg/provider.go [hunk 3]
```diff
@@ -1,213 +1,164 @@
...
-// WithPort specifies the port of the flagd server. Defaults to 8013
+// WithPort specifies the expectPort of the flagd server. Defaults to 8013
 func WithPort(port uint16) ProviderOption {
 	return func(p *Provider) {
 		p.providerConfiguration.Port = port
 	}
 }
 
-// WithHost specifies the host name of the flagd server. Defaults to localhost
+// WithHost specifies the expectHost name of the flagd server. Defaults to localhost
```
Comment (toddbaert) [inline providers/flagd/pkg/provider.go:161]:
```suggestion
// WithHost specifies the host name of the flagd server. Defaults to localhost
```
guessing this is a typo

File: providers/flagd/pkg/provider.go [hunk 4]
```diff
@@ -1,213 +1,164 @@
...
+// If certificate loading fails insecure credentials will be used instead
 func WithCertificatePath(path string) ProviderOption {
 	return func(p *Provider) {
 		p.providerConfiguration.CertificatePath = path
 		p.providerConfiguration.TLSEnabled = true
 	}
 }
 
-// WithPort specifies the port of the flagd server. Defaults to 8013
+// WithPort specifies the expectPort of the flagd server. Defaults to 8013
```
Comment (toddbaert) [inline providers/flagd/pkg/provider.go:154]:
```suggestion
// WithPort specifies the port of the flagd server. Defaults to 8013
```
guessing this is a typo

File: providers/flagd/pkg/provider.go [hunk 5]
```diff
@@ -1,213 +1,164 @@
...
-	SocketPath      string
-	TLSEnabled      bool
-}
+	logger                logr.Logger
+	providerConfiguration *providerConfiguration
+	service               IService
+	status                of.State
 
-type ProviderOption func(*Provider)
+	eventStream chan of.Event
```
Comment (toddbaert) [inline providers/flagd/pkg/provider.go:18]: channels "feel" great for this.

File: providers/flagd/pkg/service/service.go [hunk 6]
```diff
@@ -50,131 +76,292 @@ type resolutionResponseConstraints interface {
...
 
-	return res.Msg, nil
+	go func() {
+		s.startEventStream(ctx)
+	}()
+
+	s.cancelHook = cancelFunc
+
+	return nil
+}
```
Comment (toddbaert) [inline providers/flagd/pkg/service/service.go:95]: Unless I'm mistaken, we have a similar issue here as we did in Java, where we don't know for sure that the event stream is successfully connected when `Init` is complete.

Can we block here until a ready event?

File: providers/flagd/pkg/service/service.go [hunk 7]
```diff
@@ -50,131 +76,292 @@ type resolutionResponseConstraints interface {
...
+
+	if s.cache.IsEnabled() {
+		fromCache, ok := s.cache.GetCache().Get(key)
+		if ok {
+			fromCacheResDetail, ok := fromCache.(openfeature.BoolResolutionDetail)
+			if ok {
+				fromCacheResDetail.Reason = ReasonCached
+				return fromCacheResDetail
+			}
+		}
```
Comment (toddbaert) [inline providers/flagd/pkg/service/service.go:113]: It would be great if we could use generics or some other mechanism to reduce this duplication.

File: providers/flagd/pkg/service/service.go [hunk 8]
```diff
@@ -1,40 +1,66 @@
 package service
 
 import (
+	schemaConnectV1 "buf.build/gen/go/open-feature/flagd/bufbuild/connect-go/schema/v1/schemav1connect"
```
Comment (toddbaert) [inline providers/flagd/pkg/service/service.go:4]: This is a nit, but now that you've isolated most of the gRPC code here, you may consider calling this file grpc_service... not necessary though.

If you do, I would recommend you do it with `git mv ...` so that it's easier to see in review.

File: providers/flagd/README.md (hunk 2 above)
Comment (Kavindu-Dodan) [reply providers/flagd/README.md:54]: We always supported in-memory cache but lacked documentation under environment configuration section - https://github.com/open-feature/go-sdk-contrib/blob/6545b4c16db57114456c58a3eb2966ec461ae9f9/providers/flagd/README.md#using-flagdfromenv

File: providers/flagd/pkg/provider.go (hunk 3 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/provider.go:161]: 🤦 this is from a test variable renaming. Thank you for noticing

File: providers/flagd/pkg/provider.go (hunk 4 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/provider.go:154]: yup, this is a rename ALL mistake 

original usage - https://github.com/open-feature/go-sdk-contrib/blob/b777a3d9efe858be0eacf26056d31cf9debec444/providers/flagd/pkg/provider_test.go#L13-L14

File: providers/flagd/pkg/provider.go (hunk 5 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/provider.go:18]: Yeah, emit the event and let SDK handle it. IMO this is much better than shared objects + locks

File: providers/flagd/pkg/service/service.go (hunk 8 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:4]: Good suggestion. But I would like to defer this renaming to a follow-up PR. This one already has quite a lot of changes (IMO too many for a single pr 😞 ) so changing the packaging could confuse you even more.

File: providers/flagd/pkg/service/service.go (hunk 6 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:95]: 🤦 this is something I always forget about the spec.

I fixed this with the latest commit - https://github.com/open-feature/go-sdk-contrib/pull/317/commits/607ba6e84fcf4b2c278d65a13706ba3eba6518da 

however, this also shows there's some corner cases when it comes to handling events 🤔 something we could improve

File: providers/flagd/pkg/service/service.go (hunk 7 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:113]: +1 on this. I did not attempt this to reduce review efforts. Change in the PR fully maps to current main's logic. But once merged, we might be able to find some commons and extract them to generic handlers

File: providers/flagd/pkg/configuration.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
+package flagd
+
+import (
+	"fmt"
+	"github.com/go-logr/logr"
+	"github.com/open-feature/go-sdk-contrib/providers/flagd/pkg/cache"
```
Comment (thisthat) [inline providers/flagd/pkg/configuration.go:6]: I think these are not formatted correctly

File: providers/flagd/pkg/configuration.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+		EventStreamConnectionMaxAttempts: defaultMaxEventStreamRetries,
+		Host:                             defaultHost,
+		MaxCacheSize:                     defaultMaxCacheSize,
+		Port:                             defaultPort,
+		TLSEnabled:                       defaultTLS,
+	}
+}
+
+// updateFromEnvVar is a utility to update configurations based on current environment variables
+func (cfg *providerConfiguration) updateFromEnvVar(logger logr.Logger) {
```
Comment (thisthat) [inline providers/flagd/pkg/configuration.go:56]: I see that the logger is not part of `providerConfiguration` but it feels weird to have it as an argument of a function since it could vary between calls to this function.

File: providers/flagd/pkg/configuration.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+				))
+		} else {
+			cfg.Port = uint16(port)
+		}
+	}
+
+	host := os.Getenv(flagdHostEnvironmentVariableName)
+	if host != "" {
+		cfg.Host = host
+	}
```
Comment (thisthat) [inline providers/flagd/pkg/configuration.go:74]: nit. This way we scope `host` to the bare minimum
```suggestion
	if host := os.Getenv(flagdHostEnvironmentVariableName); host != "" {
		cfg.Host = host
	}
```

File: providers/flagd/pkg/configuration.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+	socketPath := os.Getenv(flagdSocketPathEnvironmentVariableName)
+	if socketPath != "" {
+		cfg.SocketPath = socketPath
+	}
+
+	certificatePath := os.Getenv(flagdServerCertPathEnvironmentVariableName)
+	if certificatePath != "" || os.Getenv(flagdTLSEnvironmentVariableName) == "true" {
```
Comment (thisthat) [inline providers/flagd/pkg/configuration.go:82]: maybe worth trimming and lowercase the value of `flagdTLSEnvironmentVariableName`

File: providers/flagd/pkg/service/service.go [hunk 13]
```diff
@@ -193,66 +380,173 @@ func handleError(err error) openfeature.ResolutionError {
...
-			delay = s.baseRetryDelay // reset delay if the connection was successful before failing
-		}
-		if err != nil && i <= maxAttempts {
+func (s *Service) EventChannel() <-chan of.Event {
+	return s.events
+}
+
+// startEventStream - starts listening to flagd event stream with retries.
+// If retrying is exhausted, an event with openfeature.ProviderError is emitted.
+func (s *Service) startEventStream(ctx context.Context) {
```
Comment (thisthat) [inline providers/flagd/pkg/service/service.go:389]: maybe worth adding a line saying that this function is expected to be called in a go routine

File: providers/flagd/pkg/service/service.go [hunk 14]
```diff
@@ -193,66 +380,173 @@ func handleError(err error) openfeature.ResolutionError {
...
+	var delay = s.baseRetryDelay
+	var attempts = 1
+
+	// wraps connection with retry attempts
+	for attempts <= s.maxRetries {
+		s.logger.V(logger.Debug).Info("connecting to event stream")
+		err := s.streamClient(ctx)
```
Comment (thisthat) [inline providers/flagd/pkg/service/service.go:396]: I think here we missed resetting the connection attempts if we successfully connect.

File: providers/flagd/pkg/service/service.go (hunk 6 above)
Comment (toddbaert) [reply providers/flagd/pkg/service/service.go:95]: Yep, that looks good to me.

File: providers/flagd/pkg/configuration.go (hunk 10 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/configuration.go]: Updated with the commit - https://github.com/open-feature/go-sdk-contrib/pull/317/commits/158f552c75623a69b753f67afb835e14b4ddbeaf

I made logger part of the config instance creation so that we don't have to call it with a logger

File: providers/flagd/pkg/configuration.go (hunk 11 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/configuration.go]: Updated all if conditions to use the same format

File: providers/flagd/pkg/service/service.go (hunk 14 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:396]: This handled at the else block [^1]

And error handling the stream means there's something wrong, which we should retry. This could be a connection loss, packet errors or something else.

However, if there's no error, which can happen with a provider shutdown, stream end, we reset attempts and starts fresh connection establishment attempts 

[^1]: https://github.com/open-feature/go-sdk-contrib/blob/470ee860a64ccade08b6e764962e5b9d50c0a210/providers/flagd/pkg/service/service.go#L397-L409

File: providers/flagd/pkg/service/service.go (hunk 13 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:390]: Good suggestion. Done - https://github.com/open-feature/go-sdk-contrib/pull/317/commits/3eee6b23362073325def650e01a5973d831bae74

File: providers/flagd/pkg/configuration.go (hunk 9 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/configuration.go:6]: We do not enforfce gofumpt here. I have run intellij formatter and this is the output it gives 🤔  I think this is simple alphabetic ordering

File: providers/flagd/pkg/configuration.go (hunk 12 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/configuration.go]: not sure this is needed 🤔  I didn't want to add/change the behavior of this logic.

File: providers/flagd/pkg/configuration.go (hunk 9 above)
Comment (thisthat) [reply providers/flagd/pkg/configuration.go:6]: you are right 👍

File: providers/flagd/pkg/service/service.go (hunk 14 above)
Comment (thisthat) [reply providers/flagd/pkg/service/service.go:397]: Maybe I am misread, but I think `stream.Receive()` exits only when the stream ends. Hence, if I fail two times to connect, at the third attempt I can connect, after a while the connection drops, the retry starts at 3 instead of 0 🤔

File: providers/flagd/pkg/configuration.go (hunk 12 above)
Comment (thisthat) [reply providers/flagd/pkg/configuration.go]: Agree, this doesn't need to be part of this PR. However, maybe it is worth unifying across languages how to process bool env var. @toddbaert do we have somewhere in the spec a common way for SDKs to handle truthy and falsy values?

File: providers/flagd/pkg/service/service.go (hunk 14 above)
Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:397]: Yup, that's something that the current logic doesn't handle. I will improve this

Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:391]: Improvement done with this commit - https://github.com/open-feature/go-sdk-contrib/pull/317/commits/44870f92ea07f9923f2ef3f47acf7ca500aad2c4 

Retry counter and sleep delay now live in its own structure

File: providers/flagd/pkg/service/service.go [hunk 15]
```diff
@@ -432,6 +421,9 @@ func (s *Service) streamClient(ctx context.Context) error {
 	s.logger.V(logger.Info).Info("connected to event stream")
```
Comment (thisthat) [inline providers/flagd/pkg/service/service.go:422]: Based on the docs (see [Creating Providers](https://github.com/open-feature/flagd/blob/main/docs/other_resources/creating_providers.md#provider-lifecycle-initialization-and-shutdown)), I think here we need to know if we previously failed to emit a `Ready` and `Changed` events

Comment (Kavindu-Dodan) [reply providers/flagd/pkg/service/service.go:422]: The initial ready state is emitted from flagd. This happens at the first event stream connection establishment. 

State change : NOT_READY -> READY   (this event is emitted by SDK, flagd provider do not emit this)

From there on, we can get Config changes, Provider shutdown or connection errors , which are handled by the `streamClient ` logic. 

I do not see the need to store the previous state and could rely on current stream message to produce the valid state

**Review by thisthat — APPROVED** [review]
Thanks @Kavindu-Dodan 🙌

**Review by toddbaert — APPROVED** [review]

File: providers/flagd/README.md
Comment (beeme1mr) [inline providers/flagd/README.md:1]: Please add a "Supported Events" section to the readme. I should look similar to the one available for JS.

https://github.com/open-feature/js-sdk-contrib/tree/main/libs/providers/flagd#supported-events

Comment (Kavindu-Dodan) [reply providers/flagd/README.md:1]: Added with the latest commit. See preview - https://github.com/open-feature/go-sdk-contrib/blob/16fe631890b3ec5ab1eb304b12f9da71628b32ef/providers/flagd/README.md#supported-events

**Review by beeme1mr — APPROVED** [review]
